packages/amplify-prompts/src/progressbars/spinner.ts

```diff
@@ -3,64 +3,83 @@
  * and use a re writable block instead.
  */
 
-import { AmplifyTerminal as Terminal, StringObj } from './terminal';
+import { AmplifyTerminal, AmplifyTerminal as Terminal, StringObj } from './terminal';
 
 /**
  * Amplify spinner instance
  */
 export class AmplifySpinner {
-    private frameCount : number;
-    private frames : string[];
-    private timer!: ReturnType<typeof setTimeout>;
-    private prefixText : string;
-    private terminal: Terminal;
-    private refreshRate: number;
+  private frameCount: number;
+  private frames: string[];
+  private timer!: ReturnType<typeof setTimeout>;
+  private prefixText: string;
+  private terminal: Terminal | null;
+  private refreshRate: number;
 
-    constructor(text : string) {
-      this.frameCount = 0;
-      this.frames = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
-      this.prefixText = text;
-      this.terminal = new Terminal();
-      this.refreshRate = 50;
-    }
+  constructor() {
+    this.frameCount = 0;
+    this.frames = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
+    this.prefixText = '';
+    this.refreshRate = 50;
+    this.terminal = null;
+  }
 
-    /**
-     * Render function
-     */
-    render() : void {
-      if (this.timer) {
-        clearTimeout(this.timer);
-      }
-      const lines = [{
-        renderString: `${this.frames[this.frameCount]} ${this.prefixText}`,
-        color: '',
-      }];
-      this.frameCount = ++this.frameCount % this.frames.length;
+  /**
+   * Render function
+   */
+  private render(): void {
+    if (this.timer) {
+      clearTimeout(this.timer);
+    }
+    const lines = [{
+      renderString: `${this.frames[this.frameCount]} ${this.prefixText}`,
+      color: '',
+    }];
+    this.frameCount = ++this.frameCount % this.frames.length;
+    if (this.terminal) {
       this.terminal.writeLines(lines);
```

Comment:
does it make sense to execute other logic in this method / class if there's no terminal ?
Should this whole thing just be no-op (aka NullObject) if there's no terminal?